### ISSUE TITLE
StructureData conversion from ase/pymatgen too slow

### ISSUE THREAD

**Issue by ltalirz** [aiidateam/aiida_core#595]
The conversion of pymatgen/ASE objects containing structures with several hundreds of atoms in to StructureData is ridiculously slow.
In my test (see also attached output from cProfile: 4 structures with ~800 atoms each, ASE + pymatgen), about 4s per are spent per structure on this rather trivial conversion procedure. Note that this strongly outweighs the time spent by ASE or pymatgen on actually parsing the .cif files and generating their representation in python objects.

The attached profiles illustrate two problems:

1. [append_atom](https://github.com/aiidateam/aiida_core/blob/develop/aiida/orm/data/structure.py#L1195) is called anew for every atom in the structure. This clearly adds overhead, but could be tolerated if the overhead isn't too large (will still have to investigate this).
2. Every call to append_atom actually spends about 90% inside [set_attr](https://github.com/aiidateam/aiida_core/blob/develop/aiida/orm/data/__init__.py#L95), i.e. updating some python-representation of node objects. As far as I can see, this is completely unnecessary and should happen only once at the very end, when the StructureData object is complete.

I am creating this issue as a reference, as I will be working on this in a fork on my github account. 

@fernandogargiulo1986 mentioned that one might take inspiration from the ParameterData object, which apparently has different modes of construction.

Further ideas & comments are of course highly appreciated!

[cprofile.zip](https://github.com/aiidateam/aiida_core/files/1080978/cprofile.zip)

**Comment by giovannipizzi** [comment]
Are you trying on Django or sqlalchemy?
is this maybe due to the new function `clean_value` inside `orm/implementation/general/node.py`?

**Comment by ltalirz** [comment]
1) I've been using Django for the moment. I guess I should switch, right?
2) No idea... I'll investigate

**Comment by giovannipizzi** [comment]
No, it's ok, in any case we should investigate both (but it's possible/we expect that sqlalchemy is both faster and occupies less space in the DB).

If the time is spent before calling `.store()`, anyway, it _should_ be backend-independent.

**Comment by ltalirz** [comment]
Yes, I think you're right. What almost 100% of the `_set_attr` call boils down to in the end is actually `deepcopy`.

**Comment by ltalirz** [comment]
I was on the `restapi` branch before since I was trying out Materials cloud.
Now that I switched to `develop` it has gotten slower again by ~50%. And indeed, now 95% of the time spent in `set_ase` is taken by `clean_value`.
(see attached cProfile output)
[cprofile.zip](https://github.com/aiidateam/aiida_core/files/1087513/cprofile.zip)

**Comment by ltalirz** [comment]
Ok, two more pieces of information:

(1) If I play stupid and simply deactivate clean_values (by always returning `value` directly), I don't receive any error messages and the time spent in the conversion procedure from ASE objects to AiiDA objects is reduced by approx. a factor of 5. This is still not very good; now it takes roughly the same time as for ASE to parse the .cif files.
However, using the django backend, parsing+conversion together now takes approximately as much time as storing the information in the database.

(2) simply running clean_value on a numpy array of size 1 million takes ~6s, so the `is_instance` checks actually aren't necessarily *that* slow. Also when operating on the actual structures in aiida, the cost per `isinstance` call is just ~1e-6s. The problem is that there are >3M `isinstance` calls per .cif file (which contain a few hundred atoms).
I think this is what we need to tackle.

**Comment by broeder-j** [comment]
might be connected to issue #267, therefore please also check if get_formula becomes faster and close the issue #267 if it does.

**Comment by ltalirz** [comment]
Finally getting back to this.

The root of the problem is in `StructureData.append_site`:
https://github.com/aiidateam/aiida_core/blob/e376e68674b7decb490c6792f28e05939573c647/aiida/orm/data/structure.py#L1199-L1202
For every atom that is appended, it uses `_set_attr` on the whole list of sites, causing the (recursive) calls to `clean_value` to be O(N^2) in the number of sites N.

Simply replacing this line by
`self.get_attr('sites', []).append(new_site.get_raw())`
for my example reduces the number of calls to `clean_value` from 4.0M to 576 and the time spent inside `set_ase` from 27.3s to 1.2s

Yet, this this is obviously a hack as I am updating the node object directly, not via `_set_attr`.

For a cleaner solution there are multiple possibilities and it isn't clear to me, what is the best way forward.
One way could be to forward the information that only part of the a new attribute needs to be cleaned further down the line (either to `_set_attr` or even down to `clean_value`).
Another way could be to provide an off switch for `clean_value` and give methods like `append_site` the possibility to handle the cleaning internally "on their own risk".

@giovannipizzi @DropD What do you think?

**Comment by ltalirz** [comment]
#733 fixes the problem with `clean_value` - the speed of handling the python objects is now on the same level as the time for parsing the .cif files.

For the design of a new, light-weight structuredata class, we should open a new issue.
